src/lib/tuning/abstract_selector.hpp

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "tuning/tuning_choice.hpp"
+#include "tuning/tuning_operation.hpp"
+
+namespace opossum {
+
+/**
+ * An AbstractSelector transforms an unordered list of TuningChoices into
+ * a concrete sequence of TuningOperations.
+ *
+ * It considers a cost budget that the entire operation sequence as well as any
+ * continuous subsequence from the beginning must not exceed.
+ * The operation sequence is prioritized by expected performance impact on the system,
+ * so that the most beneficial operations come before less useful operations.
+ *
+ * The underlying problem relates to the knapsack problem, but goes beyond that
+ * in scope (e.g. choices invalidating other choices).
+ * For more information, see e.g. Wikipedia:
+ *   https://en.wikipedia.org/wiki/Knapsack_problem
+ */
+class AbstractSelector {
+ public:
+  virtual ~AbstractSelector() {}
+  /**
+   * Determine the index operation sequence as specified above based on the
+   * given list of choices and the budget value.
+   */
+  virtual std::vector<std::shared_ptr<TuningOperation>> select(
+      const std::vector<std::shared_ptr<TuningChoice>>& choices, float budget) = 0;
```

Comment:
How do you like TuningPossibilities?

Reply:
We prefer TuningChoice: This is due to the fact that the TuningChoice is based on one specific system state, where it might already be present or not. Therefore rejecting/undoing a Choice might also result in a nontrivial TuningOperation. This aspect seems not to be represented by a term like possibility, which suggests something that might only be added to the system state.

Reply:
@Bensk1 @mrks  as discussed please decide

Reply:
https://english.stackexchange.com/questions/120475/what-is-the-difference-in-meaning-between-choice-and-option

"Options are the things and choices are our decision. Options are fixed and choices aren't."

Not really authoritative, but this is why I don't like "choice". For me, it sounds like a made decision.

Reply:
Also: http://portlandenglish.edu/blog/difference-possibility-choice-option/

Reply:
Renamed it to `TuningOption`